primitives/src/merkle_tree/mod.rs

```diff
@@ -291,14 +304,37 @@ pub trait UniversalMerkleTreeScheme: MerkleTreeScheme {
     /// Update the leaf value at a given position
     /// * `pos` - zero-based index of the leaf in the tree
     /// * `elem` - newly updated element
-    /// * `returns` - Ok(elem) if the update is success, and `elem` is the
-    ///   original element at the given `pos`. Err() if the update fails.
+    /// * `returns` - Err() if any error occurs internally. Ok(result) if the
+    ///   update is success or the given leaf is not in memory.
     fn update(
         &mut self,
         pos: impl Borrow<Self::Index>,
         elem: impl Borrow<Self::Element>,
     ) -> Result<LookupResult<Self::Element, (), ()>, PrimitivesError>;
 
+    /// Remove a leaf at the given position
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `returns` - Err() if any error occurs internally. Ok(result) if the
+    ///   update is success or the given leaf is not in memory.
+    fn remove(
+        &mut self,
+        pos: impl Borrow<Self::Index>,
+    ) -> Result<LookupResult<Self::Element, (), ()>, PrimitivesError>;
+
+    /// Apply an update function `f` at a given position
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `f` - the update function, `None` means the given leaf doesn't exist
+    ///   or should be removed.
+    /// * `returns` - Err() if any error occurs internally. Ok(result) if the
+    ///   update is success or the given leaf is not in memory.
+    fn update_with<F>(
+        &mut self,
+        pos: impl Borrow<Self::Index>,
+        f: F,
+    ) -> Result<LookupResult<Self::Element, (), ()>, PrimitivesError>
+    where
+        F: FnOnce(Option<&Self::Element>) -> Option<Self::Element>;
+
```

Comment:
why do we assume the update function only takes the old value? would it be safer to assume that it's a function of lots many other tree info, 
e.g. I only increment the fifth slot if we have more than 10 leaves with odd value. (i admit, a contrived logic)

my point being, why restrict ourselves to this API, why couldn't we just accept the `new_value: Self::Element` and let the caller figure out the derivation of the new value?

Reply:
ah, right, my mistake, anonymous functions can `move` contextual variable into the body, I forgot.